### ISSUE TITLE
Performance issue when enableMaxDepthChecks

### ISSUE THREAD

**Issue by aelfannir** [schmittjoh/serializer#1347]
| Q                | A
| ---------------- | -----
| Bug report?      | yes/no
| Feature request? | no
| BC Break report? | yes/no
| RFC?             | yes/no

I'm trying to serialize 1000 user entities, in a dev environment. 
- With `enableMaxDepthChecks` : ~ 50 sec
- Without `enableMaxDepthChecks` : ~ 1 sec

```php
public function serializer($entities)
{
    $context = new SerializationContext();
    $context
        ->enableMaxDepthChecks()
        ->setGroups(['DEFAULT'])
    ;
    $result = $this->serializer->serialize($entities, 'json', $context);

    return json_decode($result);
}
```
* Seems the issue comes from the 2nd expression in this condition 
https://github.com/schmittjoh/serializer/blob/a682e33e8d3811c8b7a6c5b2f92e05855f69edd5/src/Exclusion/DepthExclusionStrategy.php#L39-L41

**Comment by aelfannir** [comment]
Moved to JMSSerializerBundle 
https://github.com/schmittjoh/JMSSerializerBundle/issues/871

**Comment by aelfannir** [comment]
re-open

**Comment by goetas** [comment]
Is it just slow/CPU time or are there more SQL queries executed?

Are you using the `@Depth()` exclusion?

**Comment by aelfannir** [comment]
> Is it just slow/CPU time

No
<br>
> or are there more SQL queries executed

I'm not sure  <br>

> Are you using the @Depth() exclusion?

Yes, I use `@MaxDepth` exclusion.

#### Here's the code

* User.php
```php
<?php
declare(strict_types=1);

namespace App\Entity;

use App\Repository\UserRepository;
use App\Traits\FileEntity;
use App\Traits\IdEntity;
use App\Traits\TimestampEntity;
use App\Traits\UserEntity;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Security\Core\User\UserInterface;
use JMS\Serializer\Annotation as JMS;
use Vich\UploaderBundle\Mapping\Annotation as Vich;
use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;

/**
 * https://github.com/dustin10/VichUploaderBundle/issues/1226
 *
 * @Vich\Uploadable
 */
#[
    JMS\ExclusionPolicy(policy: JMS\ExclusionPolicy::NONE),
    ORM\Entity(repositoryClass: UserRepository::class),
    ORM\Table(name: 'users'),
    ORM\UniqueConstraint(name: 'user_unique', columns: ['username']),
    UniqueEntity(fields: 'username', message: 'VALIDATION.SERVER.UNIQUE_ENTITY'),
]
class User implements UserInterface
{
    use IdEntity;
    use FileEntity;
    use UserEntity;
    use TimestampEntity;

    /**
     * https://www.doctrine-project.org/projects/doctrine-orm/en/2.10/reference/attributes-reference.html
     *
     * @ORM\JoinColumns({
     *   @ORM\JoinColumn(onDelete="SET NULL")
     * })
     */
    #[
        JMS\Groups(['default']),
        JMS\Type(Role::class),
        JMS\MaxDepth(1),
        ORM\ManyToOne(targetEntity: Role::class, inversedBy: 'users')
    ]
    private ?Role $role;

    public function getRoles()
    {
        return [$this->getRole()->getRoleKey()];
    }

    public function getRole(): ?Role
    {
        return $this->role;
    }

    public function setRole(?Role $role): self
    {
        $this->role = $role;

        return $this;
    }
}
```

* Role.php
```php
<?php
declare(strict_types=1);

namespace App\Entity;

use App\Repository\RoleRepository;
use App\Traits\IdEntity;
use App\Traits\PriorityEntity;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use JMS\Serializer\Annotation as JMS;
use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;
use Symfony\Component\Validator\Constraints as Assert;

#[
    JMS\ExclusionPolicy(policy: JMS\ExclusionPolicy::NONE),
    ORM\Entity(repositoryClass: RoleRepository::class),
    ORM\Table(name: 'roles'),
    UniqueEntity(fields: 'nameFr', message: 'VALIDATION.SERVER.UNIQUE_ENTITY'),
]
class Role
{
    use IdEntity;
    use PriorityEntity;

    public const SUPER_ADMIN = 'ROLE_SUPER_ADMIN';
    public const ADMIN = 'ROLE_ADMIN';
    public const CUSTOM = 'ROLE_CUSTOM';

    public const ROLES = [
        self::SUPER_ADMIN,
        self::ADMIN,
        self::CUSTOM,
    ];

    public function __construct()
    {
        $this->routes = new ArrayCollection();
        $this->users = new ArrayCollection();
    }

    #[
        JMS\Groups(['DEFAULT']),
        JMS\Type('string'),
        ORM\Column,
        Assert\Choice(choices: self::ROLES, message: 'VALIDATION.SERVER.UNIQUE_ENTITY')
    ]
    private string $roleKey = self::CUSTOM;

    #[
        Assert\NotBlank(message: 'VALIDATION.SERVER.NOT_BLANK'),
        Assert\NotNull(message: 'VALIDATION.SERVER.NOT_NULL'),
        JMS\Groups(['DEFAULT']),
        JMS\Type('string'),
        ORM\Column(unique: true)
    ]
    private ?string $nameFr;

    /**
     * @ JMS\Accessor(setter="syncRoutes")
     */
    #[
        JMS\Groups(['DEFAULT']),
        JMS\MaxDepth(2),
        JMS\Type('ArrayCollection<App\Entity\Route>'),
        ORM\ManyToMany(targetEntity: Route::class, inversedBy: 'roles'),
        ORM\JoinTable(name: 'role_route')
    ]
    private $routes;

    /**
     * @ORM\OneToMany(targetEntity="App\Entity\User", mappedBy="role", cascade={"persist"})
     */
    #[
        JMS\Groups([]),
        JMS\MaxDepth(1),
        JMS\Type('ArrayCollection<App\Entity\User>'),
        JMS\Accessor(setter: 'addUsers'),
        ORM\OneToMany(mappedBy: 'role', targetEntity: User::class, cascade: ['persist']),
    ]
    private $users;

    public function setRoleKey(string $roleKey): self
    {
        $this->roleKey = $roleKey;

        return $this;
    }

    public function getNameFr(): ?string
    {
        return $this->nameFr;
    }

    public function setNameFr(string $nameFr): self
    {
        $this->nameFr = $nameFr;

        return $this;
    }

    public function getRoleKey(): ?string
    {
        return $this->roleKey;
    }

    /**
     * @return Collection
     */
    public function getUsers(): Collection
    {
        return $this->users;
    }

    public function addUser(User $user): self
    {
        if (!$this->users->contains($user)) {
            $this->users[] = $user;
            $user->setRole($this);
        }

        return $this;
    }

    public function removeUser(User $user): self
    {
        if ($this->users->removeElement($user)) {
            // set the owning side to null (unless already changed)
            if ($user->getRole() === $this) {
                $user->setRole(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection
     */
    public function getRoutes(): Collection
    {
        return $this->routes;
    }

    public function addRoute(Route $route): self
    {
        if (!$this->routes->contains($route)) {
            $this->routes[] = $route;
        }

        return $this;
    }

    public function syncRoutes($routes): self
    {
        $this->getRoutes()->clear();
        foreach ($routes as $route) {
            $this->addRoute($route);
        }

        return $this;
    }

    public function removeRoute(Route $route): self
    {
        $this->routes->removeElement($route);

        return $this;
    }
}
```

* Route.php
```php
<?php
declare(strict_types=1);

namespace App\Entity;

use App\Repository\RouteRepository;
use App\Traits\IdEntity;
use App\Traits\PriorityEntity;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use JMS\Serializer\Annotation as JMS;
use Symfony\Component\Validator\Constraints as Assert;

/**
 * @JMS\ExclusionPolicy("none")
 * @ORM\Entity(repositoryClass=RouteRepository::class)
 * @ORM\Table(name="routes")
 */
class Route
{
    use IdEntity;
    use PriorityEntity;

    public const ROLES_NEW = 'ROLES.NEW';
    public const ROLES_EDIT = 'ROLES.EDIT';
    public const ROLES_DELETE = 'ROLES.DELETE';
    public const ROLES_LIST = 'ROLES.LIST';

    public const USERS_NEW = 'USERS.NEW';
    public const USERS_EDIT = 'USERS.EDIT';
    public const USERS_DELETE = 'USERS.DELETE';
    public const USERS_LIST = 'USERS.LIST';
    public const USER_PROFILE_EDIT = 'USER.PROFILE.EDIT';
    public const USER_PASSWORD_EDIT = 'USER.PASSWORD.EDIT';

    public const ROUTES = [
        self::USER_PROFILE_EDIT, self::USER_PASSWORD_EDIT,
        self::USERS_NEW, self::USERS_EDIT, self::USERS_DELETE, self::USERS_LIST,
        self::ROLES_NEW, self::ROLES_EDIT, self::ROLES_DELETE, self::ROLES_LIST,
    ];

    public function __construct()
    {
        $this->roles = new ArrayCollection();
    }

    /**
     * @Assert\NotBlank(message="VALIDATION.SERVER.NOT_BLANK")
     * @Assert\NotNull(message="VALIDATION.SERVER.NOT_NULL")
     * @JMS\Groups({"DEFAULT"})
     * @JMS\Type("string")
     * @ORM\Column(unique=true)
     */
    private ?string $routeKey;

    /**
     * @JMS\Groups({})
     * @JMS\Type("ArrayCollection<App\Entity\Role>")
     * @JMS\MaxDepth(2)
     * @ORM\ManyToMany(targetEntity="App\Entity\Role", mappedBy="routes")
     */
    private ArrayCollection $roles;

    /**
     * @JMS\VirtualProperty
     * @JMS\SerializedName("_roles")
     */
    public function getRolesKeys(): array
    {
        $keys = [];
        foreach ($this->getRoles() as $role) {
            $keys[] = $role->getRoleKey();
        }

        return $keys;
    }

    public function getRouteKey(): ?string
    {
        return $this->routeKey;
    }

    public function setRouteKey(string $routeKey): self
    {
        $this->routeKey = $routeKey;

        return $this;
    }

    /**
     * @return Collection
     */
    public function getRoles(): Collection
    {
        return $this->roles;
    }

    public function addRole(Role $role): self
    {
        if (!$this->roles->contains($role)) {
            $this->roles[] = $role;
            $role->addRoute($this);
        }

        return $this;
    }

    public function removeRole(Role $role): self
    {
        if ($this->roles->removeElement($role)) {
            $role->removeRoute($this);
        }

        return $this;
    }
}
```

* jms_serrializer.yaml
```
jms_serializer:
  property_naming:
    id: 'jms_serializer.identical_property_naming_strategy'
```

* services.yaml
```yaml
services:
    ...
    jms_serializer.object_constructor:
        class: App\Service\JMSObjectConstructor
        arguments:
            - '@doctrine'
            - '@jms_serializer.unserialize_object_constructor'
    jms_serializer.serialized_name_annotation_strategy:
        class: JMS\Serializer\Naming\SerializedNameAnnotationStrategy
        arguments:
            - '@jms_serializer.identical_property_naming_strategy'
 ```
 * JMSObjectConstructor.php
 ```php
 <?php

namespace App\Service;

use Doctrine\Persistence\ManagerRegistry;
use JMS\Serializer\Visitor\DeserializationVisitorInterface;
use JMS\Serializer\Metadata\ClassMetadata;
use JMS\Serializer\DeserializationContext;
use JMS\Serializer\Construction\ObjectConstructorInterface;

class JMSObjectConstructor implements ObjectConstructorInterface
{
    /**
     * constructor.
     * @param ManagerRegistry $managerRegistry
     * @param ObjectConstructorInterface $fallbackConstructor
     */
    public function __construct(
        protected ManagerRegistry $managerRegistry,
        protected ObjectConstructorInterface $fallbackConstructor
    )
    { }

    /**
     * {@inheritdoc}
     */
    public function construct(DeserializationVisitorInterface $visitor, ClassMetadata $metadata, $data, array $type, DeserializationContext $context ): ?object
    {
        // Locate possible ObjectManager
        $objectManager = $this->managerRegistry->getManagerForClass($metadata->name);

        if ( ! $objectManager) {
//            dump("1"); exit();
            // No ObjectManager found, proceed with normal deserialization
            return $this->fallbackConstructor->construct($visitor, $metadata, $data, $type, $context);
        }

        // Locate possible ClassMetadata
        $classMetadataFactory = $objectManager->getMetadataFactory();

        if ($classMetadataFactory->isTransient($metadata->name)) {
            // No ClassMetadata found, proceed with normal deserialization
//            dump("2"); exit();
            return $this->fallbackConstructor->construct($visitor, $metadata, $data, $type, $context);
        }

        if( ! $data) return null;

        // Managed entity, check for proxy load
        if ( ! is_array($data)) {
            // Single identifier, load proxy
//            dump("3",$data); exit();
            return $objectManager->getReference($metadata->name, $data);
        }

        // Fallback to default constructor if missing identifier(s)
        $classMetadata  = $objectManager->getClassMetadata($metadata->name);
        $identifierList = array();

        foreach ($classMetadata->getIdentifierFieldNames() as $name) {
            if ( ! array_key_exists($name, $data)) {
//                dump("4"); exit();
                /* Todo test
                Manytoone : create, update
                Manytomany : create, update
                todo
                */
                $classname = $metadata->name;
                return new $classname();
//                dump($data);
//                $objectManager->initializeObject($data);

                return $this->fallbackConstructor->construct($visitor, $metadata, $data, $type, $context);
            }

            $identifierList[$name] = $data[$name];
        }

        // Entity update, load it from database
        $object = $objectManager->find($metadata->name, $identifierList);

        $objectManager->initializeObject($object);

//        dump("object",$object);
//        $classname = $metadata->name;
//        return new $classname();
        return $object;
    }
}
```

**Comment by scyzoryck** [comment]
Hey @aelfannir ! 
I've create a simple fix for the issue in PR #1369 Could you take a look if it will help in your case? :) According to benchmarks serialisation with enabled max depth exclusions are still a bit slower compared to disabled checks, but there difference should be much lower than now.

**Comment by aelfannir** [comment]
Hi @scyzoryck, thank you for the time.

I've reproduced the same use case, and test it again with and without `enableMaxDepthChecks`, i comfirm PR #1369 fixes the performance issue !
